src/main/java/htsjdk/variant/vcf/VCFMetaDataLines.java

```diff
@@ -0,0 +1,298 @@
+    // Namespace key used for "other" (unstructured, non-ID) metadata lines. This string needs to be different from
+    // any string in the set of legal structured line types in knownStructuredLineKeys.
+    private static String OTHER_KEY = "OTHER";
+
+    /**
+     * Add all metadata lines from Set. If a duplicate line is encountered (same key/ID pair for
+     * structured lines, or duplicate content for unstructured lines with identical keys)
+     * only the first line will be retained.
+     *
+     * @param newMetaData Set of lines to be added to the list.
+     * @throws IllegalArgumentException if a fileformat line is added
+     */
+    public void addAllMetaDataLines(Set<VCFHeaderLine> newMetaData) {
+        newMetaData.stream().forEach(hl -> addMetaDataLine(hl));
+    }
+
+    /**
+     * Add a metadata line to the list. If a duplicate line is encountered (same key/ID pair for
+     * structured lines, or duplicate content for unstructured lines with identical keys)
+     * only the first line will be retained.
+     *
+     * @param headerLine header line to attempt to add
+     * @throws IllegalArgumentException if a fileformat line is added
+     */
+    public void addMetaDataLine(final VCFHeaderLine headerLine) {
+        String key = makeKeyForLine(headerLine);
+        if ( mMetaData.get(key) != null ) {
+            final String message = String.format(
+                    "Attempt to add header line (%s) collides with existing line header line (%s). " +
+                            "The existing line will be retained",
+                    mMetaData.get(key),
+                    headerLine);
+            //TODO: should this also detect/reject multiple "assembly" or "reference" lines ?
+            if (VCFHeaderVersion.isFormatString(headerLine.getKey())) { // Throw if there is more than one fileformat line
+                throw new TribbleException(message);
+            } else if (VCFUtils.getVerboseVCFLogging()) {
+                // TODO: The previous header implementation would round trip lines with duplicate IDs by preserving
+                // TODO: them in the master header line list maintained by VCFHeader, but would silently drop them
+                // TODO: from the typed header line lists, so the duplicates would not surface in queries (i.e. via
+                // TODO: getInfoHeaderLines()).
+                //
+                // TODO: This implementation doesn't allow the duplicates to be preserved (the 4.2/4.3 specs expressly
+                // TODO: forbid it), so they're dropped here. There are several Picard and GATK test files around with
+                // TODO: duplicate ID lines, which will have to be modified in order for the tests to pass.
+                //
+                // TODO: Is it sufficient to log a warning for this case (duplicate key) ? Should we add code
+                // TODO: here to throw only for v4.3+ ? Only if VCFUtils.getStrictVCFVersionValidation is set ? Or both ?
```

Comment:
After reading through the whole PR, I start to think that VCF requires a `ValidationStringency` enum similar to SAM/BAM parsing to use in different parts, with STRICT value by default (maybe SILENT until the 3 PRs are in to keep the actual behaviour).